site/src/lib/middleware/hasValidApiKey.middleware.ts

```diff
@@ -0,0 +1,34 @@
+import { Middleware } from "next-connect";
+import { BaseApiRequest, BaseApiResponse } from "../handler/baseHandler";
+
+import { ApiKey } from "../model/apiKey.model";
+import { formatErrors } from "../../util/api";
+
+export const hasValidApiKeyMiddleware: Middleware<
+  BaseApiRequest,
+  BaseApiResponse
+> = async (req, res, next) => {
+  try {
+    const apiKey = req.headers["x-api-key"];
+    if (!apiKey || typeof apiKey !== "string") {
+      throw new Error(
+        "A valid API key must be provided in a 'x-api-key' header.",
+      );
+    }
+
+    const { db } = req;
+
+    await ApiKey.validate(db, { apiKeyString: apiKey });
+
+    next();
+  } catch (err) {
+    return res.status(401).send(
```

Comment:
Different code styles (`throw new Error` and `res.status().send()) made me think that the responses would be shaped differently. They are the same, which is good for the end users:

![Screenshot 2022-01-18 at 20 07 02](https://user-images.githubusercontent.com/608862/150010729-0be38c04-6c5d-4cf8-9bb6-7bf867db1389.png)

I wonder if we can align the implementation or at least state why we cannot do this. I originally thought that `throw new Error` would give us `5xx` – that’s why I went to check this.

Reply:
The error is immediately caught by the catch block - I threw it like this to make sure that this error is handled the same as anything thrown by `ApiKey.validate`.

I can instead repeat the `res.status(401).send`, although this makes it more likely they will diverge - is this clearer/better?

```typescript
  try {
    const apiKey = req.headers["x-api-key"];
    if (!apiKey || typeof apiKey !== "string") {
      return res
        .status(401)
        .send(
          formatErrors({
            msg: "A valid API key must be provided in a 'x-api-key' header.",
          }),
        );
    }
    ...
  } catch (err) {
    return res.status(401).send(
      formatErrors({
        msg:
          err instanceof Error
            ? err.message
            : "Unknown error validating API key.",
      }),
    );
  }
```

Reply:
I've gone ahead and done this - let me know if it's better

Reply:
Yeah that looks a bit more coherent thanks! 👍